src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1635,9 +1635,19 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     debug!("place_element_conflict: DISJOINT-OR-EQ-LOCAL");
                     Overlap::EqualOrDisjoint
                 } else {
-                    // different locals - base case, disjoint
-                    debug!("place_element_conflict: DISJOINT-LOCAL");
-                    Overlap::Disjoint
+                    let ty = self.mir.local_decls[*l1].ty;
+                    match ty.sty {
+                        ty::TyAdt(def, _) if def.is_union() => {
+                            // different locals from a union
+                            debug!("place_element_conflict: STUCK-UNION-LOCAL");
+                            Overlap::Arbitrary
+                        },
+                        _ => {
+                            // different locals - base case, disjoint
+                            debug!("place_element_conflict: DISJOINT-LOCAL");
+                            Overlap::Disjoint
+                        },
+                    }
```

Comment:
I think this is not quite right. Local variables should always be disjoint from one another. For example, borrowing `x` while `y` is borrowed should never be an error, no matter the type of `x`.